core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1244,6 +1244,66 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
 
+  /**
+    * Like [[Stream.flatMap]] but interrupts the inner stream when new elements arrive in the outer stream.
+    *
+    * The implementation will try to preserve chunks like [[Stream.merge]].
+    *
+    * Finializers of each inner stream are guaranteed to run before the next inner stream starts.
+    *
+    * When the outer stream stops gracefully, the currently running inner stream will continue to run.
+    *
+    * When an inner stream terminates/interrupts, nothing happens until the next element arrives
+    * in the outer stream(i.e the outer stream holds the stream open during this time or else the
+    * stream terminates)
+    *
+    * When either the inner or outer stream fails, the entire stream fails.
+    *
+    */
+  def switchMap[F2[x] >: F[x], O2](f: O => Stream[F2, O2])(
+      implicit F2: Concurrent[F2]): Stream[F2, O2] =
+    Stream.eval(async.unboundedQueue[F2, Option[Either[Throwable, Chunk[O2]]]]).flatMap {
+      downstream =>
+        Stream.eval(async.unboundedQueue[F2, Option[(O, Deferred[F2, Unit])]]).flatMap { next =>
+          Stream.eval(Ref.of[F2, Option[Deferred[F2, Unit]]](None)).flatMap { haltRef =>
+            // consumes the upstream by sending a packet(stream and a ref) to the next queue and
+            // also does error handling/inturruption/switching by terminating the currently running
+            // inner stream
+            def runUpstream =
+              this
+                .onFinalize(next.enqueue1(None))
+                .evalMap { x =>
+                  Deferred[F2, Unit].flatMap(halt =>
+                    haltRef.modify {
+                      case None       => halt.some -> F2.unit
+                      case Some(last) => halt.some -> last.complete(())
+                    }.flatten *> next.enqueue1((x -> halt).some))
+                }
+                .handleErrorWith(e =>
+                  Stream.eval(downstream.enqueue1(Left(e).some) *>
+                    haltRef.modify(x => None -> x.fold(F2.unit)(_.complete(()))).flatten))
+
+            // consumes the packets produced by runUpstream and drains the content to downstream
+            // upon error/interruption, the downstream will also be signalled for termination
+            def runInner =
+              next.dequeue.unNoneTerminate
+                .onFinalize(downstream.enqueue1(None))
+                .flatMap {
+                  case (o, haltInner) =>
+                    f(o).chunks
+                      .evalMap(x => downstream.enqueue1(Right(x).some))
+                      .interruptWhen(haltInner.get.attempt)
+                      .handleErrorWith(e => Stream.eval(downstream.enqueue1(Left(e).some)))
+                }
+
+            // continuously dequeue the downstream while we process the upstream
+            downstream.dequeue.unNoneTerminate.rethrow
+              .flatMap(Stream.chunk(_))
+              .concurrently(runInner.merge(runUpstream))
```

Comment:
So something like `concurrently(runInner.concurrently(runUpstream))`? Or do we want a parJoin of some sort?

Reply:
I'd try with the `concurrently(... concurrently`, no queues are allocated. `parJoin` allocates one
Do test it though 😄